File: packages/ios/OpenCode/API/Connection.swift
import Foundation

/// How the app reaches the OpenCode server.
enum ConnectionMode: String, Codable, Sendable {
    /// Through the OpenCode Remote relay (server is behind NAT).
    case relay
    /// Straight to a reachable OpenCode server (LAN / dev).
    case direct
}

/// Everything needed to connect, persisted in the Keychain (it holds a token).
struct ConnectionConfig: Codable, Equatable, Sendable {
    var mode: ConnectionMode = .relay

    // relay mode
    var relayURL = ""
    var tunnelID = ""
    var token = ""

    // direct mode
    var directURL = ""
    var password: String?

    /// The base URL the API client targets. For relay mode this is the
    /// per-tunnel proxy path on the relay; for direct mode it's the server URL.
    var baseURL: String {
        switch mode {
        case .relay:
            return relayURL.trimmedSlashes + "/t/" + tunnelID
        case .direct:
            return directURL.trimmedSlashes
        }
    }

    /// Whether the user has filled in enough to attempt a connection.
    var isComplete: Bool {
        switch mode {
        case .relay:
            return !relayURL.isEmpty && !tunnelID.isEmpty && !token.isEmpty
        case .direct:
            return !directURL.isEmpty
        }
    }

    /// Parses a pairing payload produced by the connector, accepting either an
    /// `opencode://pair?relay=…&tunnel=…&token=…` URL (also used for QR codes
    /// and deep links) or a bare `relay=…&tunnel=…&token=…` query string.
    static func fromPairing(_ raw: String) -> ConnectionConfig? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }

        var query = trimmed
        if let comps = URLComponents(string: trimmed), comps.scheme != nil {
            guard comps.host == "pair" || comps.path.contains("pair") else { return nil }
            query = comps.percentEncodedQuery ?? ""
        }

        var items: [String: String] = [:]
        for pair in query.split(separator: "&") {
            let kv = pair.split(separator: "=", maxSplits: 1)
            guard kv.count == 2 else { continue }
            let key = String(kv[0])
            let value = String(kv[1]).removingPercentEncoding ?? String(kv[1])
            items[key] = value
        }

        guard let relay = items["relay"], let tunnel = items["tunnel"], let token = items["token"],
              !relay.isEmpty, !tunnel.isEmpty, !token.isEmpty
        else { return nil }

        return ConnectionConfig(mode: .relay, relayURL: relay, tunnelID: tunnel, token: token)
    }
}

extension String {
    /// Trims surrounding whitespace and any trailing slashes.
    var trimmedSlashes: String {
        var s = trimmingCharacters(in: .whitespacesAndNewlines)
        while s.hasSuffix("/") { s.removeLast() }
        return s
    }
}

File: packages/android/app/src/main/java/studio/eugenezakharov/opencode/api/Connection.kt
package studio.eugenezakharov.opencode.api

import android.net.Uri

/** How the app reaches the OpenCode server. */
enum class ConnectionMode {
    /** Through the OpenCode Remote relay (server is behind NAT). */
    RELAY,

    /** Straight to a reachable OpenCode server (LAN / dev). */
    DIRECT,
}

/**
 * Everything needed to connect. Mirrors the iOS `ConnectionConfig`.
 *
 * For relay mode the base URL is the per-tunnel proxy path on the relay
 * (`relayURL/t/{tunnelID}`); for direct mode it is the server URL.
 */
data class ConnectionConfig(
    val mode: ConnectionMode = ConnectionMode.RELAY,
    // relay mode
    val relayURL: String = "",
    val tunnelID: String = "",
    val token: String = "",
    // direct mode
    val directURL: String = "",
    val password: String? = null,
) {
    val baseURL: String
        get() = when (mode) {
            ConnectionMode.RELAY -> relayURL.trimmedSlashes() + "/t/" + tunnelID
            ConnectionMode.DIRECT -> directURL.trimmedSlashes()
        }

    /** Whether the user has filled in enough to attempt a connection. */
    val isComplete: Boolean
        get() = when (mode) {
            ConnectionMode.RELAY ->
                relayURL.isNotEmpty() && tunnelID.isNotEmpty() && token.isNotEmpty()
            ConnectionMode.DIRECT -> directURL.isNotEmpty()
        }

    companion object {
        /**
         * Parses a pairing payload produced by the connector, accepting either an
         * `opencode://pair?relay=…&tunnel=…&token=…` URL (also used for QR codes
         * and deep links) or a bare `relay=…&tunnel=…&token=…` query string.
         */
        fun fromPairing(raw: String): ConnectionConfig? {
            val trimmed = raw.trim()
            if (trimmed.isEmpty()) return null

            var query = trimmed
            val uri = runCatching { Uri.parse(trimmed) }.getOrNull()
            if (uri != null && uri.scheme != null) {
                val isPair = uri.host == "pair" || (uri.path?.contains("pair") == true)
                if (!isPair) return null
                query = uri.encodedQuery ?: ""
            }

            val items = mutableMapOf<String, String>()
            for (pair in query.split("&")) {
                val kv = pair.split("=", limit = 2)
                if (kv.size != 2) continue
                val value = Uri.decode(kv[1])
                items[kv[0]] = value
            }

            val relay = items["relay"]
            val tunnel = items["tunnel"]
            val token = items["token"]
            if (relay.isNullOrEmpty() || tunnel.isNullOrEmpty() || token.isNullOrEmpty()) {
                return null
            }

            return ConnectionConfig(
                mode = ConnectionMode.RELAY,
                relayURL = relay,
                tunnelID = tunnel,
                token = token,
            )
        }
    }
}

/** Trims surrounding whitespace and any trailing slashes. */
fun String.trimmedSlashes(): String {
    var s = trim()
    while (s.endsWith("/")) s = s.dropLast(1)
    return s
}
